File: src/student/student.rs
use chrono::{DateTime, Local};
use std::collections::HashMap;
use uuid::Uuid;

pub struct StudentDb<'a> {
    pub students: HashMap<Uuid, StudentData<'a>>,
}

pub struct StudentData<'a> {
    pub student_id: Uuid,
    pub data: Student<'a>,
    pub status: StudentStatus,
    pub created_at: DateTime<Local>,
}

pub struct Student<'a> {
    pub fist_name: &'a str,
    pub last_name: &'a str,
    pub dob: Dob,
    pub gender: Gender,
    pub email: &'a str,
    pub program: Program,
}

pub struct Dob {
    pub day: u16,
    pub month: Month,
    pub year: u32,
}

pub enum Month {
    Jenuary,
    February,
    March,
    April,
    May,
    June,
    July,
    August,
    September,
    October,
    November,
    December,
}

#[derive(Debug, PartialEq)]
pub enum Gender {
    Male,
    Female,
}

#[derive(Debug, PartialEq)]
pub enum Program {
    Mathematics,
    History,
    Economics,
    Literature,
    Physics,
    Biochemistry,
    ComputerScience,
    Architecture,
}

#[derive(Debug, PartialEq)]
pub enum StudentStatus {
    Active,
    Expelled,
    Graduated,
}

File: frontend/types/student.ts
export interface Student {
  id?: number;
  student_id: string; // Unique student identifier
  first_name: string;
  last_name: string;
  dob: string; // ISO date string
  gender: Gender;
  email: string;
  program: Program;
  student_status: StudentStatus;
  created_at?: string;
  updated_at?: string;
}

export enum Gender {
  MALE = "MALE",
  FEMALE = "FEMALE",
  OTHER = "OTHER",
}

export enum Program {
  COMPUTER_SCIENCE = "COMPUTER_SCIENCE",
  ENGINEERING = "ENGINEERING",
  BUSINESS = "BUSINESS",
  MEDICINE = "MEDICINE",
  LAW = "LAW",
  ARTS = "ARTS",
  SCIENCE = "SCIENCE",
}

export enum StudentStatus {
  ACTIVE = "ACTIVE",
  INACTIVE = "INACTIVE",
  GRADUATED = "GRADUATED",
  SUSPENDED = "SUSPENDED",
  TRANSFERRED = "TRANSFERRED",
}

export interface CreateStudentRequest {
  first_name: string;
  last_name: string;
  dob: string;
  gender: Gender;
  email: string;
  program: Program;
  student_status?: StudentStatus;
}

export interface UpdateStudentRequest extends CreateStudentRequest {
  id: number;
  student_id: string;
}

export interface StudentFilters {
  program?: Program;
  gender?: Gender;
  status?: StudentStatus;
  search?: string;
}
